File: packages/sdk-go/offline.go
package ubag

import (
	"fmt"
	"sync"
	"time"
)

type OfflineEntry struct {
	ID         string         `json:"id"`
	Request    map[string]any `json:"request"`
	EnqueuedAt string         `json:"enqueued_at"`
	Attempts   int            `json:"attempts"`
}

type OfflineStore interface {
	Read() []OfflineEntry
	Write(entries []OfflineEntry)
}

type MemoryOfflineStore struct {
	mu      sync.Mutex
	entries []OfflineEntry
}

func NewMemoryOfflineStore() *MemoryOfflineStore { return &MemoryOfflineStore{} }

func (m *MemoryOfflineStore) Read() []OfflineEntry {
	m.mu.Lock()
	defer m.mu.Unlock()
	out := make([]OfflineEntry, len(m.entries))
	copy(out, m.entries)
	return out
}

func (m *MemoryOfflineStore) Write(entries []OfflineEntry) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.entries = entries
}

type OfflineQueue struct {
	store   OfflineStore
	counter int
}

func NewOfflineQueue(store OfflineStore) *OfflineQueue { return &OfflineQueue{store: store} }

func (q *OfflineQueue) Enqueue(request map[string]any) OfflineEntry {
	entries := q.store.Read()
	q.counter++
	entry := OfflineEntry{
		ID:         fmt.Sprintf("q_%d_%d", time.Now().UnixNano(), q.counter),
		Request:    request,
		EnqueuedAt: time.Now().UTC().Format(time.RFC3339Nano),
	}
	entries = append(entries, entry)
	q.store.Write(entries)
	return entry
}

// Flush sends entries FIFO; on the first sender error it persists the
// remaining entries and returns the error.
func (q *OfflineQueue) Flush(sender func(map[string]any) error) error {
	entries := q.store.Read()
	for len(entries) > 0 {
		if err := sender(entries[0].Request); err != nil {
			entries[0].Attempts++
			q.store.Write(entries)
			return err
		}
		entries = entries[1:]
		q.store.Write(entries)
	}
	return nil
}

func (q *OfflineQueue) Size() int { return len(q.store.Read()) }

File: packages/sdk-rust/src/offline.rs
//! Offline queue with a pluggable store. The in-memory store is always
//! available; a sled-backed store is provided behind the `offline-sled` feature.

use serde::{Deserialize, Serialize};
use std::sync::Mutex;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct OfflineEntry {
    pub id: String,
    pub request: serde_json::Value,
    pub attempts: u32,
}

pub trait OfflineStore {
    fn read(&self) -> Vec<OfflineEntry>;
    fn write(&self, entries: Vec<OfflineEntry>);
}

#[derive(Default)]
pub struct MemoryOfflineStore {
    entries: Mutex<Vec<OfflineEntry>>,
}

impl OfflineStore for MemoryOfflineStore {
    fn read(&self) -> Vec<OfflineEntry> {
        self.entries.lock().unwrap().clone()
    }
    fn write(&self, entries: Vec<OfflineEntry>) {
        *self.entries.lock().unwrap() = entries;
    }
}

pub struct OfflineQueue<S: OfflineStore> {
    store: S,
    counter: Mutex<u64>,
}

impl<S: OfflineStore> OfflineQueue<S> {
    pub fn new(store: S) -> Self {
        OfflineQueue { store, counter: Mutex::new(0) }
    }

    pub fn enqueue(&self, request: serde_json::Value) -> OfflineEntry {
        let mut entries = self.store.read();
        let mut c = self.counter.lock().unwrap();
        *c += 1;
        let entry = OfflineEntry { id: format!("q_{}", *c), request, attempts: 0 };
        entries.push(entry.clone());
        self.store.write(entries);
        entry
    }

    /// Sends entries FIFO; on first sender error persists remaining and returns Err.
    pub fn flush<F>(&self, mut sender: F) -> Result<(), String>
    where
        F: FnMut(&serde_json::Value) -> Result<(), String>,
    {
        let mut entries = self.store.read();
        while !entries.is_empty() {
            match sender(&entries[0].request) {
                Ok(()) => {
                    entries.remove(0);
                    self.store.write(entries.clone());
                }
                Err(e) => {
                    entries[0].attempts += 1;
                    self.store.write(entries);
                    return Err(e);
                }
            }
        }
        Ok(())
    }

    pub fn size(&self) -> usize {
        self.store.read().len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trip() {
        let q = OfflineQueue::new(MemoryOfflineStore::default());
        q.enqueue(serde_json::json!({"target": "mock"}));
        q.enqueue(serde_json::json!({"target": "mock"}));
        assert_eq!(q.size(), 2);
        let mut sent = 0;
        q.flush(|_| { sent += 1; Ok(()) }).unwrap();
        assert_eq!(sent, 2);
        assert_eq!(q.size(), 0);
    }

    #[test]
    fn retains_on_error() {
        let q = OfflineQueue::new(MemoryOfflineStore::default());
        q.enqueue(serde_json::json!({"target": "mock"}));
        let r = q.flush(|_| Err("offline".to_string()));
        assert!(r.is_err());
        assert_eq!(q.size(), 1);
    }
}
